internal/controller/scheduler/scheduler.go

```diff
@@ -243,56 +251,68 @@ func (w *jobWrapper) Build(skipCheckout bool) (*batchv1.Job, error) {
 	kjob.Spec.TTLSecondsAfterFinished = &ttl
 
 	podSpec := &kjob.Spec.Template.Spec
+
+	containerEnv := append([]corev1.EnvVar{}, env...)
+	containerEnv = append(containerEnv, []corev1.EnvVar{
+		{
+			Name:  "BUILDKITE_AGENT_EXPERIMENT",
+			Value: "kubernetes-exec",
+		},
+		{
+			Name:  "BUILDKITE_BOOTSTRAP_PHASES",
+			Value: "plugin,command",
+		},
+		{
+			Name:  "BUILDKITE_AGENT_NAME",
+			Value: "buildkite",
+		},
+		{
+			Name:  "BUILDKITE_PLUGINS_PATH",
+			Value: "/tmp",
+		},
+		{
+			Name:  clicommand.RedactedVars.EnvVar,
+			Value: strings.Join(redactedVars, ","),
+		},
+		{
+			Name:  "BUILDKITE_SHELL",
+			Value: "/bin/sh -ec",
+		},
+		{
+			Name:  "BUILDKITE_ARTIFACT_PATHS",
+			Value: w.envMap["BUILDKITE_ARTIFACT_PATHS"],
+		},
+		{
+			Name:  "BUILDKITE_SOCKETS_PATH",
+			Value: "/workspace/sockets",
+		},
+	}...)
+
 	for i, c := range podSpec.Containers {
-		command := strings.Join(append(c.Command, c.Args...), " ")
+		// If the command is empty, use the command from the step
+		command := w.job.Command
+		if len(c.Command) > 0 {
+			command = strings.Join(append(c.Command, c.Args...), " ")
+		}
 		c.Command = []string{"/workspace/buildkite-agent"}
 		c.Args = []string{"bootstrap"}
 		c.ImagePullPolicy = corev1.PullAlways
-		c.Env = append(c.Env, env...)
+		c.Env = append(c.Env, containerEnv...)
 		c.Env = append(c.Env,
 			corev1.EnvVar{
 				Name:  "BUILDKITE_COMMAND",
 				Value: command,
 			},
-			corev1.EnvVar{
-				Name:  "BUILDKITE_AGENT_EXPERIMENT",
-				Value: "kubernetes-exec",
-			},
-			corev1.EnvVar{
-				Name:  "BUILDKITE_BOOTSTRAP_PHASES",
-				Value: "plugin,command",
-			},
-			corev1.EnvVar{
-				Name:  "BUILDKITE_AGENT_NAME",
-				Value: "buildkite",
-			},
 			corev1.EnvVar{
 				Name:  "BUILDKITE_CONTAINER_ID",
 				Value: strconv.Itoa(i + systemContainerCount),
 			},
-			corev1.EnvVar{
-				Name:  "BUILDKITE_PLUGINS_PATH",
-				Value: "/tmp",
-			},
-			corev1.EnvVar{
-				Name:  "BUILDKITE_SOCKETS_PATH",
-				Value: "/workspace/sockets",
-			},
-			corev1.EnvVar{
-				Name:  clicommand.RedactedVars.EnvVar,
-				Value: strings.Join(redactedVars, ","),
-			},
-			corev1.EnvVar{
-				Name:  "BUILDKITE_SHELL",
-				Value: "/bin/sh -ec",
-			},
-			corev1.EnvVar{
-				Name:  "BUILDKITE_ARTIFACT_PATHS",
-				Value: w.envMap["BUILDKITE_ARTIFACT_PATHS"],
-			})
+		)
```

Comment:
Most of these were constant in `i` so were moved into `containerEnv` outside the loop.